gateway/utils/utils.go

```diff
@@ -22,68 +23,96 @@ func NewServerTLSConfig(ctx context.Context, certs []tls.Certificate, cquery cty
 		InsecureSkipVerify: true, // nolint: gosec
 		MinVersion:         tls.VersionTLS13,
 		VerifyPeerCertificate: func(certificates [][]byte, _ [][]*x509.Certificate) error {
-			if len(certificates) > 0 {
-				if len(certificates) != 1 {
-					return errors.Errorf("tls: invalid certificate chain")
-				}
-
-				cert, err := x509.ParseCertificate(certificates[0])
-				if err != nil {
-					return errors.Wrap(err, "tls: failed to parse certificate")
-				}
-
-				// validation
-				var owner sdk.Address
-				if owner, err = sdk.AccAddressFromBech32(cert.Subject.CommonName); err != nil {
-					return errors.Wrap(err, "tls: invalid certificate's subject common name")
-				}
-
-				// 1. CommonName in issuer and Subject must match and be as Bech32 format
-				if cert.Subject.CommonName != cert.Issuer.CommonName {
-					return errors.Wrap(err, "tls: invalid certificate's issuer common name")
-				}
-
-				// 2. serial number must be in
-				if cert.SerialNumber == nil {
-					return errors.Wrap(err, "tls: invalid certificate serial number")
-				}
-
-				// 3. look up certificate on chain
-				var resp *ctypes.QueryCertificatesResponse
-				resp, err = cquery.Certificates(
-					ctx,
-					&ctypes.QueryCertificatesRequest{
-						Filter: ctypes.CertificateFilter{
-							Owner:  owner.String(),
-							Serial: cert.SerialNumber.String(),
-							State:  "valid",
-						},
-					},
-				)
-				if err != nil {
-					return errors.Wrap(err, "tls: unable to fetch certificate from chain")
-				}
-				if (len(resp.Certificates) != 1) || !resp.Certificates[0].Certificate.IsState(ctypes.CertificateValid) {
-					return errors.New("tls: attempt to use non-existing or revoked certificate")
-				}
-
-				clientCertPool := x509.NewCertPool()
-				clientCertPool.AddCert(cert)
-
-				opts := x509.VerifyOptions{
-					Roots:                     clientCertPool,
-					CurrentTime:               time.Now(),
-					KeyUsages:                 []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
-					MaxConstraintComparisions: 0,
-				}
-
-				if _, err = cert.Verify(opts); err != nil {
-					return errors.Wrap(err, "tls: unable to verify certificate")
-				}
+			if _, err := VerifyCertChain(ctx, certificates, "", x509.ExtKeyUsageClientAuth, cquery); err != nil {
+				return err
 			}
 			return nil
 		},
 	}
 
 	return cfg, nil
 }
+
+type certChain interface {
+	*x509.Certificate | []byte
+}
+
+func VerifyCertChain[T certChain](
```

Comment:
1. do we really need generics here?
2. if Owner address is returned then better name would be `VerifyOwnerCert`. Overall issue with this approach is that some GRPC methods may not provider client cert. `Provider status` is good example. So assumption is that returned `sdk.Address` is never `nil`, which is not true in this case

Reply:
1. It depends. It buys us 1) the ability to do the length checks before type assertion and 2) provide a sensible union type to the caller so they know what to pass. We can get rid of the generics once we remove the REST gateway (it's the only code passing in `[][]byte`. Another option would be to create wrapper functions that are typed to `[][]byte` and `[]*x509.Certificate` (but this would require duplicating the length checks). I'm not a huge fan of Go's generics, but it does make this particular code DRY.
2. That's a good point, I like that name better and thanks for the context.

Reply:
Also, it occurs to me that the type constraint interface should be called `cert`, not `certChain`.

Reply:
lets remove generics from this usecase, `[][]byte` `[]*x509.Certificate` would expect that first expects valid encoded certificate. if there is need for `[][]byte` to exist wrapper would be clear and better option